File: AlgorithmsProblemSolvingJava/src/io/programminglife/youtube/DiameterBinaryTree.java
package io.programminglife.youtube;

public class DiameterBinaryTree {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        maxDepth(root);
        return ans;
    }

    public int maxDepth(TreeNode tree) {
        if (tree == null) {
            return 0;
        }

        int leftDepth = maxDepth(tree.left);
        int rightDepth = maxDepth(tree.right);

        ans = Math.max(ans, leftDepth + rightDepth);

        return Math.max(leftDepth, rightDepth) + 1;
    }
}

File: AlgorithmsProblemSolvingPython/youtube/diameter_binary_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def diameterOfBinaryTree(self, root: TreeNode) -> int:
    def maxDepth(root):
        if root is None:
            return 0
        left_height = maxDepth(root.left)
        right_height = maxDepth(root.right)
        ans[0] = max(ans[0], left_height + right_height)
        return max(left_height, right_height) + 1
    ans = [0]
    maxDepth(root)
    return ans[0]
